src/g2cinq.c

```diff
@@ -88,33 +89,53 @@ g2c_inq_msg(int g2cid, int msg_num, unsigned char *discipline, int *num_fields,
             unsigned char *local_version)
 {
     G2C_MESSAGE_INFO_T *msg;
-    /* Is this an open GRIB2 file? */
-    if (g2cid < 0 || g2cid > G2C_MAX_FILES || g2c_file[g2cid].g2cid != g2cid)
+    int ret = G2C_NOERROR;
+
+    /* Check input parameters. */
+    if (g2cid < 0 || g2cid > G2C_MAX_FILES)
         return G2C_EBADID;
+    if (msg_num < 0)
+        return G2C_EINVAL;
+
+    /* If using threading, lock the mutex. */
+    MUTEX_LOCK(m);
+
+    /* Find the open file. */
+    if (g2c_file[g2cid].g2cid != g2cid)
+        ret = G2C_EBADID;
 
     /* Find the file and message. */
-    for (msg = g2c_file[g2cid].msg; msg; msg = msg->next)
+    if (!ret)
```

Comment:
Does this all need to be in an if block? i.e., if ret!=0 could you just return ret here?

Reply:
Because of threading, the mutex must always be unlocked. So functions cannot return from the middle, they must exit from the bottom of the function, when the mutex has been unlocked.